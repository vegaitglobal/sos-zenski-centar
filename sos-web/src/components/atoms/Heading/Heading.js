import { forwardRef } from 'react';
import { StyledHeading } from './Heading.styles';

export const Heading = forwardRef(
  (
    {
      type,
      as = type,
      fontWeight,
      textTransform,
      textAlign,
      color,
      children,
      ...props
    },
    ref,
  ) => (
    <StyledHeading
      ref={ref}
      as={as}
      $type={type}
      $fontWeight={fontWeight}
      $textTransform={textTransform}
      $textAlign={textAlign}
      $color={color}
      {...props}
    >
      {children}
    </StyledHeading>
  ),
);
