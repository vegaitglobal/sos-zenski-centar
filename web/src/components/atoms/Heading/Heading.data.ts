import type { ReactNode } from 'react';
import { ColorType } from '../../../styles/config/theme';
import type { FontWeight, TextAlign, TextTransform } from '../../../styles/config/variables';

export enum HeadingType {
	H1 = 'h1',
	H2 = 'h2',
	H3 = 'h3',
	H4 = 'h4',
	H5 = 'h5',
	H6 = 'h6',
}

export interface HeadingProps {
	as?: string | React.ComponentType<unknown>;
	type?: HeadingType;
	fontWeight?: FontWeight;
	textTransform?: TextTransform;
	textAlign?: TextAlign;
	color?: ColorType;
	children: ReactNode;
}

export interface StyledHeadingProps {
	$type: HeadingType;
	$fontWeight?: FontWeight;
	$textTransform?: TextTransform;
	$textAlign?: TextAlign;
	$color?: ColorType;
}
