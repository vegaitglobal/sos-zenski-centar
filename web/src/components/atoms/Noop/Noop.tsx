import { ReactElement } from 'react';

export const Noop = (): ReactElement => <></>;
