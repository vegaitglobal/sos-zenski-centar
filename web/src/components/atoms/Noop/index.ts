export { Noop as default } from './Noop';
