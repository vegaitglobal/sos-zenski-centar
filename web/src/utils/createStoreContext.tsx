import { createContext, PropsWithChildren, FunctionComponent, Context, useContext } from 'react';
import useStoreFactory from '../hooks/useStoreFactory';

/**
 * Creates a context for the store created by the passed `storeFactory`.
 * Additionally returns the `Provider` component and the hook for the created store's context.
 *
 * @param name - name of the store
 * @param storeFactory - a function that returns an object representing the store
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const createStoreContext = <TStore extends Record<string, unknown>>(
	name: string,
	storeFactory: () => TStore,
): [FunctionComponent<unknown>, () => TStore, Context<TStore>] => {
	const context = createContext<TStore>({} as TStore); // the initial value is only to satisfy TS and won't be used

	/**
	 * Initializes and contains the context for the `TStore`.
	 */
	function StoreProvider({ children }: PropsWithChildren<unknown>): JSX.Element {
		const store = useStoreFactory(storeFactory);
		return <context.Provider value={store}>{children}</context.Provider>;
	}

	const useStore = (): TStore => {
		const store = useContext(context);
		if (store) {
			return store;
		}
		throw new ReferenceError(
			`${name} is not initialized, make sure that your component is wrapped in the 'Provider' of the store.`,
		);
	};

	return [StoreProvider, useStore, context];
};

export default createStoreContext;
