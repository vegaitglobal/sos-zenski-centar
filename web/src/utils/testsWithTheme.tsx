import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { ReactElement, ComponentType, ReactNode } from 'react';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType = ({ children }) => componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
