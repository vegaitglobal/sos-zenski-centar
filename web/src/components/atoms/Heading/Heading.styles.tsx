import styled from 'styled-components';
import { typeStyle } from '../../../styles/config/typeStyles';
import { StyledHeadingProps } from './Heading.data';

export const StyledHeading = styled.h1<StyledHeadingProps>`
	${({ $type }) => typeStyle[$type]};
	${({ $fontWeight }) => $fontWeight && `font-weight:  ${$fontWeight}`};
	${({ $textTransform }) => $textTransform && `text-transform:  ${$textTransform}`};
	${({ $textAlign }) => $textAlign && `text-align: ${$textAlign}`};
	${({ theme, $color }) => $color && `color:  ${theme.color[$color]}`};
`;
