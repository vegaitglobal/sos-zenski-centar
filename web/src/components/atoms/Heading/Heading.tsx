import { forwardRef } from 'react';
import { HeadingProps, HeadingType } from './Heading.data';
import { StyledHeading } from './Heading.styles';

export const Heading = forwardRef<HTMLHeadingElement, HeadingProps>(
	(
		{
			type = HeadingType.H1,
			as = type,
			fontWeight,
			textTransform,
			textAlign,
			color,
			children,
			...props
		},
		ref,
	) => (
		<StyledHeading
			data-testid="Heading"
			ref={ref}
			as={as}
			$type={type}
			$fontWeight={fontWeight}
			$textTransform={textTransform}
			$textAlign={textAlign}
			$color={color}
			{...props}
		>
			{children}
		</StyledHeading>
	),
);
