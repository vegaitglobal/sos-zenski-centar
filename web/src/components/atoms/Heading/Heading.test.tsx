import renderer from 'react-test-renderer';
import Noop from '../Noop';
import Heading from './index';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import { HeadingType } from './Heading.data';

describe('Heading', () => {
	describe('Render', () => {
		it('Renders Heading', () => {
			const { getByTestId } = renderWithTheme(
				<Heading>
					<Noop />
				</Heading>,
			);

			const element = getByTestId('Heading');

			expect(element).toBeTruthy();
		});

		it('Renders H1', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H1}>Heading H1</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h1');
		});

		it('Renders H2', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H2}>Heading H2</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h2');
		});

		it('Renders H3', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H3}>Heading H3</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h3');
		});

		it('Renders H4', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H4}>Heading H4</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h4');
		});

		it('Renders H5', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H5}>Heading H5</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h5');
		});

		it('Renders H6', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H6}>Heading H6</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h6');
		});

		it('Renders H3 with H2 styles', () => {
			const { getByTestId } = renderWithTheme(
				<Heading as={HeadingType.H3} type={HeadingType.H2}>
					Heading H3 with H2 styles
				</Heading>,
			);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h3');
		});

		it('Renders correctly', () => {
			const tree = renderer.create(<Heading>Heading</Heading>).toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('Has valid content', () => {
			const { getByTestId } = renderWithTheme(<Heading>Heading text</Heading>);

			const element = getByTestId('Heading');

			expect(element).toHaveTextContent('Heading text');
		});

		it('Can have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Heading>
					<span>H</span>eading
				</Heading>,
			);

			const element = getByTestId('Heading');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
