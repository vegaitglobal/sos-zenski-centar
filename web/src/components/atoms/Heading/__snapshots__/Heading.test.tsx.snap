// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Render Renders correctly 1`] = `
<h1
  className="sc-bdnxRM jCXjBt"
  data-testid="Heading"
>
  Heading
</h1>
`;
